File: product-of-array-except-self_238.py
nums = [1, 2, 3, 4]


def productExceptSelf(nums):
    if len(nums) == 0:
        return []

    result = [1] * (len(nums))  # result = [1, 1, 1, 1]

    prefix = 1
    for i, num in enumerate(nums):
        result[i] = prefix  # result = [1, 1, 2, 6]
        prefix *= num

    postfix = 1
    for i in range(len(nums) - 1, -1, -1):  # (stop, start, step)
        result[i] *= postfix  # result = [24, 12, 8, 6]
        postfix *= nums[i]

    return result


output = productExceptSelf(nums)
print(output)


# Time Complexity -> O(n)
# Space Complexity -> O(1)

File: product-of-array-except-self_238.js
/**
 * @param {number[]} nums
 * @return {number[]}
 * 
 * Time Complexity -> O(n)
 * Space Complexity -> O(n)
 */
const productExceptSelf = (nums) => {
  const n = nums.length;
  if (n === 0) return [];

  const result = new Array(n);

  // Calculate the product of elements to the left of each element
  let lefProduct = 1;
  for (let i = 0; i < n; i++) {
    result[i] = lefProduct;
    lefProduct *= nums[i];
  }

  // Calculate the product of elements to the right of each element and update the result
  let rightProduct = 1;
  for (let i = n - 1; i >= 0; i--) {
    result[i] *= rightProduct;
    rightProduct *= nums[i];
  }

  return result;
};

console.log(productExceptSelf([1, 2, 3, 4]));
